util/winutil/svcdiag_windows.go

```diff
@@ -0,0 +1,297 @@
+	}
+
+	deps := []*mgr.Service{rootSvc}
+	seen := make(map[string]struct{})
+
+	for err == nil && len(deps) > 0 {
+		err = func() error {
+			curSvc := deps[0]
+			defer curSvc.Close()
+
+			deps = deps[1:]
+
+			seen[curSvc.Name] = struct{}{}
+
+			curCfg, err := curSvc.Config()
+			if err != nil {
+				return fmt.Errorf("failed retrieving Config for service %q: %w", curSvc.Name, err)
+			}
+
+			callback(curSvc, curCfg)
+
+			for _, depName := range curCfg.Dependencies {
+				if _, ok := seen[depName]; ok {
+					continue
+				}
+
+				depSvc, err := openServiceForRead(scm, depName)
+				if err != nil {
+					return fmt.Errorf("failed opening service %q: %w", depName, err)
+				}
+
+				deps = append(deps, depSvc)
+			}
+
+			return nil
+		}()
+	}
+
+	// On error, deps might be non-empty. Its services are still open and must be closed.
+	for _, dep := range deps {
+		dep.Close()
+	}
+
+	return err
+}
+
+type svcStateLogEntry struct {
+	ServiceName string                 `json:"serviceName"`
+	ServiceType string                 `json:"serviceType"`
+	State       string                 `json:"state"`
+	StartupType string                 `json:"startupType"`
+	Triggers    *_SERVICE_TRIGGER_INFO `json:"triggers,omitempty"`
+}
+
+type _SERVICE_TRIGGER_SPECIFIC_DATA_ITEM struct {
+	dataType uint32
+	cbData   uint32
+	data     *byte
+}
+
+type serviceTriggerSpecificDataItemJSONMarshal struct {
+	DataType uint32 `json:"dataType"`
+	Data     string `json:"data,omitempty"`
+}
+
+func (tsdi *_SERVICE_TRIGGER_SPECIFIC_DATA_ITEM) MarshalJSON() ([]byte, error) {
+	m := serviceTriggerSpecificDataItemJSONMarshal{DataType: tsdi.dataType}
+
+	const maxDataLen = 128
+	data := unsafe.Slice(tsdi.data, tsdi.cbData)
+	if len(data) > maxDataLen {
+		// Only output the first maxDataLen bytes.
+		m.Data = fmt.Sprintf("%s... (truncated %d bytes)", hex.EncodeToString(data[:maxDataLen]), len(data)-maxDataLen)
+	} else {
+		m.Data = hex.EncodeToString(data)
+	}
+
+	return json.Marshal(m)
+}
+
+type _SERVICE_TRIGGER struct {
+	triggerType    uint32
+	action         uint32
+	triggerSubtype *windows.GUID
+	cDataItems     uint32
+	pDataItems     *_SERVICE_TRIGGER_SPECIFIC_DATA_ITEM
+}
+
+type serviceTriggerJSONMarshal struct {
+	TriggerType    uint32                                `json:"triggerType"`
+	Action         uint32                                `json:"action"`
+	TriggerSubtype string                                `json:"triggerSubtype,omitempty"`
+	DataItems      []_SERVICE_TRIGGER_SPECIFIC_DATA_ITEM `json:"dataItems"`
+}
+
+func (ti *_SERVICE_TRIGGER) MarshalJSON() ([]byte, error) {
+	m := serviceTriggerJSONMarshal{
+		TriggerType: ti.triggerType,
+		Action:      ti.action,
+		DataItems:   unsafe.Slice(ti.pDataItems, ti.cDataItems),
+	}
+	if ti.triggerSubtype != nil {
+		m.TriggerSubtype = ti.triggerSubtype.String()
+	}
+	return json.Marshal(m)
+}
+
+type _SERVICE_TRIGGER_INFO struct {
+	cTriggers uint32
+	pTriggers *_SERVICE_TRIGGER
+	_         *byte
```

Comment:
Could you add a comment explaining why we have this? e.g.
```suggestion
_/* pReserved */ *byte
```